pgfinder/pgio.py

```diff
@@ -23,6 +24,29 @@
 LOGGER = logging.getLogger(LOGGER_NAME)
 
 
+# FIXME: Needs type and docstring updated to dict
+def mass_libraries() -> list:
+    """
+    Loads built-in mass libraries from csv files as a dictionary.
+
+    `.csv` is chosen for consistency of data storage over the project.
+
+    Returns
+    -------
+    list
+        A list of dicts coupling mass libraries with metadata like species name and complexity
+    """
+    mass_lib_dir = Path(sys.modules[__package__].__file__).parent / "masses"
+    mass_libs = pd.read_csv(mass_lib_dir / "index.csv").groupby("Species")
+    return {
```

Comment:
I agree this could at the very least use some better comments! I'm also very open to better ways of doing things, but I found the problem a little awkward and this was pretty close to the best I could do. The end goal is a dictionary that maps species name to a list of mass libraries and their associated metadata (which is another dictionary).

The `groupby("Species")` is the key to at once enumerating the species and separating them from each other in the dataframe, then the `to_dict` call converts each row into a dictionary (a bit like `csv.DictReader()`, I think). The `**ml` bit takes that dictionary of metadata and adds the `"Content"` field with the actual, loaded mass library.

The final datastructure, pretty-printed, looks like this:
```py
{   'Clostridium difficile': [   {   'Content': b'Structure,Monoisotopicmass\ng'
                                                b'(-Ac)m|0,456.195525\ng(-Ac)m~'
                                                b'gm|0,934.375400\ng(-Ac)m~g(-A'
                                                b'c)m|0,892.364835\ng(-Ac)m~gm~'
                                                b'gm|0,1412.555275\ng(-Ac)m~g(-'
                                                b'Ac)m~gm|0,1370.544710\ng(-Ac)'
                                                b'm~g(-Ac)m~g(-Ac)m|0,1328.534'
                                                b'145\ng(-Ac)m-A|1,527.2326'
                                                b'39\ng(-Ac)m-AE|1,656.2752'
                                                b'32\ng(-Ac)m-AEJ|1,828.360'
                                                b'024\ng(-Ac)m-AEJA|1,899.39713'
                                                b'8\ng(-Ac)m-AEJC|1,931.369'
                                                b'209\ng(-Ac)m-AEJD|1,943.38696'
                                                b'7\ng(-Ac)m-AEJE|1,957.402'
                                                b'617\ng(-Ac)m-AEJF|1,975.42843'
                                                b'8\ng(-Ac)m-AEJG|1,885.381'
                                                b'488\ng(-Ac)m-AEJH|1,965.41893'
                                                b'6\ng(-Ac)m-AEJI|1,941.444'
                                                b'088\ng(-Ac)m-AEJK|1,956.45498'
                                                b'7\ng(-Ac)m-AEJM|1,959.400'
                                                b'509\ng(-Ac)m-AEJN|1,942.40295'
                                                b'1\ng(-Ac)m-AEJP|1,925.412'
                                                b'788\ng(-Ac)m-AEJQ|1,956.41860'
                                                b'2\ng(-Ac)m-AEJR|1,984.461'
                                                b'135\ng(-Ac)m-AEJS|1,915.39205'
                                                b'2\ng(-Ac)m-AEJT|1,929.407'
                                                b'702\ng(-Ac)m-AEJV|1,927.42843'
                                                b'8\ng(-Ac)m-AEJW|1,1014.439337'
                                                b'\ng(-Ac)m-AEJY|1,991.4233'
                                                b'53\ng(-Ac)m-AEJAA|1,970.43425'
                                                b'2\ng(-Ac)m-AEJAC|1,1002.40632'
                                                b'3\ng(-Ac)m-AEJAD|1,1014.42408'
                                                b'1\ng(-Ac)m-AEJAE|1,1028.43973'
                                                b'1\ng(-Ac)m-AEJAF|1,1046.46555'
                                                b'2\ng(-Ac)m-AEJAG|1,956.418602'
                                                b'\ng(-Ac)m-AEJAH|1,1036.456050'
                                                b'\ng(-Ac)m-AEJAI|1,1012.481202'
                                                b'\ng(-Ac)m-AEJAK|1,1027.492101'
                                                b'\ng(-Ac)m-AEJAM|1,1030.437623'
                                                b'\ng(-Ac)m-AEJAN|1,1013.440065'
                                                b'\ng(-Ac)m-AEJAP|1,996.449'
                                                b'902\ng(-Ac)m-AEJAQ|1,1027.455'
                                                b'716\ng(-Ac)m-AEJAR|1,1055.498'
                                                b'249\ng(-Ac)m-AEJAS|1,986.4291'
                                                b'66\ng(-Ac)m-AEJAT|1,1000.4448'
                                                b'16\ng(-Ac)m-AEJAV|1,998.46555'
                                                b'2\ng(-Ac)m-AEJAW|1,1085.47645'
                                                b'1\ng(-Ac)m-AEJAY|1,1062.46046'
                                                b'7\n',
                                     'Description': 'Contains di- tetra- and '
                                                    'hexasaccharides, '
                                                    'g(-Ac)m-A, g(-Ac)m-AE, '
                                                    'g(-Ac)m-AEJ, g(-Ac)m-AEJX '
                                                    'and all g(-Ac)m-AEJAX '
                                                    'structures, where X can '
                                                    'be any of the 20 '
                                                    'canonical amino acids. In '
                                                    'all structures, leucine '
                                                    'is removed in favour of '
                                                    'isoleucine.',
                                     'File': 'c_diff_monomers_simple.csv',
                                     'Name': 'Simple',
                                     'Species': 'Clostridium difficile'},
                                 {   'Content': b'Structure,Monoisotopicmass\ng'
                                                b'(-Ac)m|0,456.195525\ng(-Ac)m~'
                                                b'gm|0,934.375400\ng(-Ac)m~g(-A'
                                                b'c)m|0,892.364835\ng(-Ac)m~gm~'
                                                b'gm|0,1412.555275\ng(-Ac)m~g(-'
                                                b'Ac)m~gm|0,1370.544710\ng(-Ac)'
                                                b'm~g(-Ac)m~g(-Ac)m|0,1328.534'
                                                b'145\ng(-Ac)m-A|1,527.2326'
                                                b'39\ng(-Ac)m-AE|1,656.2752'
                                                b'32\ng(-Ac)m-AEJ|1,828.360'
                                                b'024\ng(-Ac)m-AEJA|1,899.39713'
                                                b'8\ng(-Ac)m-AEJC|1,931.369'
                                                b'209\ng(-Ac)m-AEJD|1,943.38696'
                                                b'7\ng(-Ac)m-AEJE|1,957.402'
                                                b'617\ng(-Ac)m-AEJF|1,975.42843'
                                                b'8\ng(-Ac)m-AEJG|1,885.381'
                                                b'488\ng(-Ac)m-AEJH|1,965.41893'
                                                b'6\ng(-Ac)m-AEJI|1,941.444'
                                                b'088\ng(-Ac)m-AEJK|1,956.45498'
                                                b'7\ng(-Ac)m-AEJM|1,959.400'
                                                b'509\ng(-Ac)m-AEJN|1,942.40295'
                                                b'1\ng(-Ac)m-AEJP|1,925.412'
                                                b'788\ng(-Ac)m-AEJQ|1,956.41860'
                                                b'2\ng(-Ac)m-AEJR|1,984.461'
                                                b'135\ng(-Ac)m-AEJS|1,915.39205'
                                                b'2\ng(-Ac)m-AEJT|1,929.407'
                                                b'702\ng(-Ac)m-AEJV|1,927.42843'
                                                b'8\ng(-Ac)m-AEJW|1,1014.439337'
                                                b'\ng(-Ac)m-AEJY|1,991.4233'
                                                b'53\ng(-Ac)m-AEJAA|1,970.43425'
                                                b'2\ng(-Ac)m-AEJAC|1,1002.40632'
                                                b'3\ng(-Ac)m-AEJCC|1,1034.37839'
                                                b'4\ng(-Ac)m-AEJAD|1,1014.42408'
                                                b'1\ng(-Ac)m-AEJCD|1,1046.39615'
                                                b'2\ng(-Ac)m-AEJDD|1,1058.41391'
                                                b'0\ng(-Ac)m-AEJAE|1,1028.43973'
                                                b'1\ng(-Ac)m-AEJCE|1,1060.41180'
                                                b'2\ng(-Ac)m-AEJDE|1,1072.42956'
                                                b'0\ng(-Ac)m-AEJEE|1,1086.44521'
                                                b'0\ng(-Ac)m-AEJAF|1,1046.46555'
                                                b'2\ng(-Ac)m-AEJCF|1,1078.43762'
                                                b'3\ng(-Ac)m-AEJDF|1,1090.45538'
                                                b'1\ng(-Ac)m-AEJEF|1,1104.47103'
                                                b'1\ng(-Ac)m-AEJFF|1,1122.49685'
                                                b'2\ng(-Ac)m-AEJAG|1,956.418602'
                                                b'\ng(-Ac)m-AEJCG|1,988.390'
                                                b'673\ng(-Ac)m-AEJDG|1,1000.408'
                                                b'431\ng(-Ac)m-AEJEG|1,1014.424'
                                                b'081\ng(-Ac)m-AEJFG|1,1032.449'
                                                b'902\ng(-Ac)m-AEJGG|1,942.4029'
                                                b'52\ng(-Ac)m-AEJAH|1,1036.4560'
                                                b'50\ng(-Ac)m-AEJCH|1,1068.4281'
                                                b'21\ng(-Ac)m-AEJDH|1,1080.4458'
                                                b'79\ng(-Ac)m-AEJEH|1,1094.4615'
                                                b'29\ng(-Ac)m-AEJFH|1,1112.4873'
                                                b'50\ng(-Ac)m-AEJGH|1,1022.4404'
                                                b'00\ng(-Ac)m-AEJHH|1,1102.4778'
                                                b'48\ng(-Ac)m-AEJAI|1,1012.4812'
                                                b'02\ng(-Ac)m-AEJCI|1,1044.4532'
                                                b'73\ng(-Ac)m-AEJDI|1,1056.4710'
                                                b'31\ng(-Ac)m-AEJEI|1,1070.4866'
                                                b'81\ng(-Ac)m-AEJFI|1,1088.5125'
                                                b'02\ng(-Ac)m-AEJGI|1,998.46555'
                                                b'2\ng(-Ac)m-AEJHI|1,1078.50300'
                                                b'0\ng(-Ac)m-AEJII|1,1054.52815'
                                                b'2\ng(-Ac)m-AEJAK|1,1027.49210'
                                                b'1\ng(-Ac)m-AEJCK|1,1059.46417'
                                                b'2\ng(-Ac)m-AEJDK|1,1071.48193'
                                                b'0\ng(-Ac)m-AEJEK|1,1085.49758'
                                                b'0\ng(-Ac)m-AEJFK|1,1103.52340'
                                                b'1\ng(-Ac)m-AEJGK|1,1013.47645'
                                                b'1\ng(-Ac)m-AEJHK|1,1093.51389'
                                                b'9\ng(-Ac)m-AEJIK|1,1069.53905'
                                                b'1\ng(-Ac)m-AEJKK|1,1084.54995'
                                                b'0\ng(-Ac)m-AEJAM|1,1030.43762'
                                                b'3\ng(-Ac)m-AEJCM|1,1062.40969'
                                                b'4\ng(-Ac)m-AEJDM|1,1074.42745'
                                                b'2\ng(-Ac)m-AEJEM|1,1088.44310'
                                                b'2\ng(-Ac)m-AEJFM|1,1106.46892'
                                                b'3\ng(-Ac)m-AEJGM|1,1016.42197'
                                                b'3\ng(-Ac)m-AEJHM|1,1096.45942'
                                                b'1\ng(-Ac)m-AEJIM|1,1072.48457'
                                                b'3\ng(-Ac)m-AEJKM|1,1087.49547'
                                                b'2\ng(-Ac)m-AEJMM|1,1090.44099'
                                                b'4\ng(-Ac)m-AEJAN|1,1013.44006'
                                                b'5\ng(-Ac)m-AEJCN|1,1045.41213'
                                                b'6\ng(-Ac)m-AEJDN|1,1057.42989'
                                                b'4\ng(-Ac)m-AEJEN|1,1071.44554'
                                                b'4\ng(-Ac)m-AEJFN|1,1089.47136'
                                                b'5\ng(-Ac)m-AEJGN|1,999.424415'
                                                b'\ng(-Ac)m-AEJHN|1,1079.461863'
                                                b'\ng(-Ac)m-AEJIN|1,1055.487015'
                                                b'\ng(-Ac)m-AEJKN|1,1070.497914'
                                                b'\ng(-Ac)m-AEJMN|1,1073.443436'
                                                b'\ng(-Ac)m-AEJNN|1,1056.445878'
                                                b'\ng(-Ac)m-AEJAP|1,996.449'
                                                b'902\ng(-Ac)m-AEJCP|1,1028.421'
                                                b'973\ng(-Ac)m-AEJDP|1,1040.439'
                                                b'731\ng(-Ac)m-AEJEP|1,1054.455'
                                                b'381\ng(-Ac)m-AEJFP|1,1072.481'
                                                b'202\ng(-Ac)m-AEJGP|1,982.4342'
                                                b'52\ng(-Ac)m-AEJHP|1,1062.4717'
                                                b'00\ng(-Ac)m-AEJIP|1,1038.4968'
                                                b'52\ng(-Ac)m-AEJKP|1,1053.5077'
                                                b'51\ng(-Ac)m-AEJMP|1,1056.4532'
                                                b'73\ng(-Ac)m-AEJNP|1,1039.4557'
                                                b'15\ng(-Ac)m-AEJPP|1,1022.4655'
                                                b'52\ng(-Ac)m-AEJAQ|1,1027.4557'
                                                b'16\ng(-Ac)m-AEJCQ|1,1059.4277'
                                                b'87\ng(-Ac)m-AEJDQ|1,1071.4455'
                                                b'45\ng(-Ac)m-AEJEQ|1,1085.4611'
                                                b'95\ng(-Ac)m-AEJFQ|1,1103.4870'
                                                b'16\ng(-Ac)m-AEJGQ|1,1013.4400'
                                                b'66\ng(-Ac)m-AEJHQ|1,1093.4775'
                                                b'14\ng(-Ac)m-AEJIQ|1,1069.5026'
                                                b'66\ng(-Ac)m-AEJKQ|1,1084.5135'
                                                b'65\ng(-Ac)m-AEJMQ|1,1087.4590'
                                                b'87\ng(-Ac)m-AEJNQ|1,1070.4615'
                                                b'29\ng(-Ac)m-AEJPQ|1,1053.4713'
                                                b'66\ng(-Ac)m-AEJQQ|1,1084.4771'
                                                b'80\ng(-Ac)m-AEJAR|1,1055.4982'
                                                b'49\ng(-Ac)m-AEJCR|1,1087.4703'
                                                b'20\ng(-Ac)m-AEJDR|1,1099.4880'
                                                b'78\ng(-Ac)m-AEJER|1,1113.5037'
                                                b'28\ng(-Ac)m-AEJFR|1,1131.5295'
                                                b'49\ng(-Ac)m-AEJGR|1,1041.4825'
                                                b'99\ng(-Ac)m-AEJHR|1,1121.5200'
                                                b'47\ng(-Ac)m-AEJIR|1,1097.5451'
                                                b'99\ng(-Ac)m-AEJKR|1,1112.5560'
                                                b'98\ng(-Ac)m-AEJMR|1,1115.5016'
                                                b'20\ng(-Ac)m-AEJNR|1,1098.5040'
                                                b'62\ng(-Ac)m-AEJPR|1,1081.5138'
                                                b'99\ng(-Ac)m-AEJQR|1,1112.5197'
                                                b'13\ng(-Ac)m-AEJRR|1,1140.5622'
                                                b'46\ng(-Ac)m-AEJAS|1,986.42916'
                                                b'6\ng(-Ac)m-AEJCS|1,1018.40123'
                                                b'7\ng(-Ac)m-AEJDS|1,1030.41899'
                                                b'5\ng(-Ac)m-AEJES|1,1044.43464'
                                                b'5\ng(-Ac)m-AEJFS|1,1062.46046'
                                                b'6\ng(-Ac)m-AEJGS|1,972.413516'
                                                b'\ng(-Ac)m-AEJHS|1,1052.450964'
                                                b'\ng(-Ac)m-AEJIS|1,1028.476116'
                                                b'\ng(-Ac)m-AEJKS|1,1043.487015'
                                                b'\ng(-Ac)m-AEJMS|1,1046.432537'
                                                b'\ng(-Ac)m-AEJNS|1,1029.434979'
                                                b'\ng(-Ac)m-AEJPS|1,1012.444816'
                                                b'\ng(-Ac)m-AEJQS|1,1043.450630'
                                                b'\ng(-Ac)m-AEJRS|1,1071.493163'
                                                b'\ng(-Ac)m-AEJSS|1,1002.424080'
                                                b'\ng(-Ac)m-AEJAT|1,1000.444816'
                                                b'\ng(-Ac)m-AEJCT|1,1032.416887'
                                                b'\ng(-Ac)m-AEJDT|1,1044.434645'
                                                b'\ng(-Ac)m-AEJET|1,1058.450295'
                                                b'\ng(-Ac)m-AEJFT|1,1076.476116'
                                                b'\ng(-Ac)m-AEJGT|1,986.429'
                                                b'166\ng(-Ac)m-AEJHT|1,1066.466'
                                                b'614\ng(-Ac)m-AEJIT|1,1042.491'
                                                b'766\ng(-Ac)m-AEJKT|1,1057.502'
                                                b'665\ng(-Ac)m-AEJMT|1,1060.448'
                                                b'187\ng(-Ac)m-AEJNT|1,1043.450'
                                                b'629\ng(-Ac)m-AEJPT|1,1026.460'
                                                b'466\ng(-Ac)m-AEJQT|1,1057.466'
                                                b'280\ng(-Ac)m-AEJRT|1,1085.508'
                                                b'813\ng(-Ac)m-AEJST|1,1016.439'
                                                b'730\ng(-Ac)m-AEJTT|1,1030.455'
                                                b'380\ng(-Ac)m-AEJAV|1,998.4655'
                                                b'52\ng(-Ac)m-AEJCV|1,1030.4376'
                                                b'23\ng(-Ac)m-AEJDV|1,1042.4553'
                                                b'81\ng(-Ac)m-AEJEV|1,1056.4710'
                                                b'31\ng(-Ac)m-AEJFV|1,1074.4968'
                                                b'52\ng(-Ac)m-AEJGV|1,984.44990'
                                                b'2\ng(-Ac)m-AEJHV|1,1064.48735'
                                                b'0\ng(-Ac)m-AEJIV|1,1040.51250'
                                                b'2\ng(-Ac)m-AEJKV|1,1055.52340'
                                                b'1\ng(-Ac)m-AEJMV|1,1058.46892'
                                                b'3\ng(-Ac)m-AEJNV|1,1041.47136'
                                                b'5\ng(-Ac)m-AEJPV|1,1024.48120'
                                                b'2\ng(-Ac)m-AEJQV|1,1055.48701'
                                                b'6\ng(-Ac)m-AEJRV|1,1083.52954'
                                                b'9\ng(-Ac)m-AEJSV|1,1014.46046'
                                                b'6\ng(-Ac)m-AEJTV|1,1028.47611'
                                                b'6\ng(-Ac)m-AEJVV|1,1026.49685'
                                                b'2\ng(-Ac)m-AEJAW|1,1085.47645'
                                                b'1\ng(-Ac)m-AEJCW|1,1117.44852'
                                                b'2\ng(-Ac)m-AEJDW|1,1129.46628'
                                                b'0\ng(-Ac)m-AEJEW|1,1143.48193'
                                                b'0\ng(-Ac)m-AEJFW|1,1161.50775'
                                                b'1\ng(-Ac)m-AEJGW|1,1071.46080'
                                                b'1\ng(-Ac)m-AEJHW|1,1151.49824'
                                                b'9\ng(-Ac)m-AEJIW|1,1127.52340'
                                                b'1\ng(-Ac)m-AEJKW|1,1142.53430'
                                                b'0\ng(-Ac)m-AEJMW|1,1145.47982'
                                                b'2\ng(-Ac)m-AEJNW|1,1128.48226'
                                                b'4\ng(-Ac)m-AEJPW|1,1111.49210'
                                                b'1\ng(-Ac)m-AEJQW|1,1142.49791'
                                                b'5\ng(-Ac)m-AEJRW|1,1170.54044'
                                                b'8\ng(-Ac)m-AEJSW|1,1101.47136'
                                                b'5\ng(-Ac)m-AEJTW|1,1115.48701'
                                                b'5\ng(-Ac)m-AEJVW|1,1113.50775'
                                                b'1\ng(-Ac)m-AEJWW|1,1200.51865'
                                                b'0\ng(-Ac)m-AEJAY|1,1062.46046'
                                                b'7\ng(-Ac)m-AEJCY|1,1094.43253'
                                                b'8\ng(-Ac)m-AEJDY|1,1106.45029'
                                                b'6\ng(-Ac)m-AEJEY|1,1120.46594'
                                                b'6\ng(-Ac)m-AEJFY|1,1138.49176'
                                                b'7\ng(-Ac)m-AEJGY|1,1048.44481'
                                                b'7\ng(-Ac)m-AEJHY|1,1128.48226'
                                                b'5\ng(-Ac)m-AEJIY|1,1104.50741'
                                                b'7\ng(-Ac)m-AEJKY|1,1119.51831'
                                                b'6\ng(-Ac)m-AEJMY|1,1122.46383'
                                                b'8\ng(-Ac)m-AEJNY|1,1105.46628'
                                                b'0\ng(-Ac)m-AEJPY|1,1088.47611'
                                                b'7\ng(-Ac)m-AEJQY|1,1119.48193'
                                                b'1\ng(-Ac)m-AEJRY|1,1147.52446'
                                                b'4\ng(-Ac)m-AEJSY|1,1078.45538'
                                                b'1\ng(-Ac)m-AEJTY|1,1092.47103'
                                                b'1\ng(-Ac)m-AEJVY|1,1090.49176'
                                                b'7\ng(-Ac)m-AEJWY|1,1177.50266'
                                                b'6\ng(-Ac)m-AEJYY|1,1154.48668'
                                                b'2\n',
                                     'Description': 'Contains di- tetra- and '
                                                    'hexasaccharides, '
                                                    'g(-Ac)m-A, g(-Ac)m-AE, '
                                                    'g(-Ac)m-AEJ, g(-Ac)m-AEJX '
                                                    'and all gm-AEJXX '
                                                    'structures, where X can '
                                                    'be any of the 20 '
                                                    'canonical amino acids; '
                                                    'for g(-Ac)m-AEJXX '
                                                    'structures, mass '
                                                    'coincidences resulting '
                                                    'from swapping the fourth '
                                                    'and fifth stem residues '
                                                    'have been removed. In all '
                                                    'structures, leucine is '
                                                    'removed in favour of '
                                                    'isoleucine.',
                                     'File': 'c_diff_monomers_non_redundant.csv',
                                     'Name': 'Non-Redundant',
                                     'Species': 'Clostridium difficile'},
                                 {   'Content': b'Structure,Monoisotopicmass\ng'
                                                b'(-Ac)m|0,456.195525\ng(-Ac)m~'
                                                b'gm|0,934.375400\ng(-Ac)m~g(-A'
                                                b'c)m|0,892.364835\ng(-Ac)m~gm~'
                                                b'gm|0,1412.555275\ng(-Ac)m~g(-'
                                                b'Ac)m~gm|0,1370.544710\ng(-Ac)'
                                                b'm~g(-Ac)m~g(-Ac)m|0,1328.534'
                                                b'145\ng(-Ac)m-A|1,527.2326'
                                                b'39\ng(-Ac)m-AE|1,656.2752'
                                                b'32\ng(-Ac)m-AEJ|1,828.360'
                                                b'024\ng(-Ac)m-AEJA|1,899.39713'
                                                b'8\ng(-Ac)m-AEJC|1,931.369'
                                                b'209\ng(-Ac)m-AEJD|1,943.38696'
                                                b'7\ng(-Ac)m-AEJE|1,957.402'
                                                b'617\ng(-Ac)m-AEJF|1,975.42843'
                                                b'8\ng(-Ac)m-AEJG|1,885.381'
                                                b'488\ng(-Ac)m-AEJH|1,965.41893'
                                                b'6\ng(-Ac)m-AEJI|1,941.444'
                                                b'088\ng(-Ac)m-AEJK|1,956.45498'
                                                b'7\ng(-Ac)m-AEJL|1,941.444'
                                                b'088\ng(-Ac)m-AEJM|1,959.40050'
                                                b'9\ng(-Ac)m-AEJN|1,942.402'
                                                b'951\ng(-Ac)m-AEJP|1,925.41278'
                                                b'8\ng(-Ac)m-AEJQ|1,956.418'
                                                b'602\ng(-Ac)m-AEJR|1,984.46113'
                                                b'5\ng(-Ac)m-AEJS|1,915.392'
                                                b'052\ng(-Ac)m-AEJT|1,929.40770'
                                                b'2\ng(-Ac)m-AEJV|1,927.428'
                                                b'438\ng(-Ac)m-AEJW|1,1014.4393'
                                                b'37\ng(-Ac)m-AEJY|1,991.423353'
                                                b'\ng(-Ac)m-AEJAA|1,970.434'
                                                b'252\ng(-Ac)m-AEJCA|1,1002.406'
                                                b'323\ng(-Ac)m-AEJDA|1,1014.424'
                                                b'081\ng(-Ac)m-AEJEA|1,1028.439'
                                                b'731\ng(-Ac)m-AEJFA|1,1046.465'
                                                b'552\ng(-Ac)m-AEJGA|1,956.4186'
                                                b'02\ng(-Ac)m-AEJHA|1,1036.4560'
                                                b'50\ng(-Ac)m-AEJIA|1,1012.4812'
                                                b'02\ng(-Ac)m-AEJKA|1,1027.4921'
                                                b'01\ng(-Ac)m-AEJLA|1,1012.4812'
                                                b'02\ng(-Ac)m-AEJMA|1,1030.4376'
                                                b'23\ng(-Ac)m-AEJNA|1,1013.4400'
                                                b'65\ng(-Ac)m-AEJPA|1,996.44990'
                                                b'2\ng(-Ac)m-AEJQA|1,1027.45571'
                                                b'6\ng(-Ac)m-AEJRA|1,1055.49824'
                                                b'9\ng(-Ac)m-AEJSA|1,986.429166'
                                                b'\ng(-Ac)m-AEJTA|1,1000.444816'
                                                b'\ng(-Ac)m-AEJVA|1,998.465'
                                                b'552\ng(-Ac)m-AEJWA|1,1085.476'
                                                b'451\ng(-Ac)m-AEJYA|1,1062.460'
                                                b'467\ng(-Ac)m-AEJAC|1,1002.406'
                                                b'323\ng(-Ac)m-AEJCC|1,1034.378'
                                                b'394\ng(-Ac)m-AEJDC|1,1046.396'
                                                b'152\ng(-Ac)m-AEJEC|1,1060.411'
                                                b'802\ng(-Ac)m-AEJFC|1,1078.437'
                                                b'623\ng(-Ac)m-AEJGC|1,988.3906'
                                                b'73\ng(-Ac)m-AEJHC|1,1068.4281'
                                                b'21\ng(-Ac)m-AEJIC|1,1044.4532'
                                                b'73\ng(-Ac)m-AEJKC|1,1059.4641'
                                                b'72\ng(-Ac)m-AEJLC|1,1044.4532'
                                                b'73\ng(-Ac)m-AEJMC|1,1062.4096'
                                                b'94\ng(-Ac)m-AEJNC|1,1045.4121'
                                                b'36\ng(-Ac)m-AEJPC|1,1028.4219'
                                                b'73\ng(-Ac)m-AEJQC|1,1059.4277'
                                                b'87\ng(-Ac)m-AEJRC|1,1087.4703'
                                                b'20\ng(-Ac)m-AEJSC|1,1018.4012'
                                                b'37\ng(-Ac)m-AEJTC|1,1032.4168'
                                                b'87\ng(-Ac)m-AEJVC|1,1030.4376'
                                                b'23\ng(-Ac)m-AEJWC|1,1117.4485'
                                                b'22\ng(-Ac)m-AEJYC|1,1094.4325'
                                                b'38\ng(-Ac)m-AEJAD|1,1014.4240'
                                                b'81\ng(-Ac)m-AEJCD|1,1046.3961'
                                                b'52\ng(-Ac)m-AEJDD|1,1058.4139'
                                                b'10\ng(-Ac)m-AEJED|1,1072.4295'
                                                b'60\ng(-Ac)m-AEJFD|1,1090.4553'
                                                b'81\ng(-Ac)m-AEJGD|1,1000.4084'
                                                b'31\ng(-Ac)m-AEJHD|1,1080.4458'
                                                b'79\ng(-Ac)m-AEJID|1,1056.4710'
                                                b'31\ng(-Ac)m-AEJKD|1,1071.4819'
                                                b'30\ng(-Ac)m-AEJLD|1,1056.4710'
                                                b'31\ng(-Ac)m-AEJMD|1,1074.4274'
                                                b'52\ng(-Ac)m-AEJND|1,1057.4298'
                                                b'94\ng(-Ac)m-AEJPD|1,1040.4397'
                                                b'31\ng(-Ac)m-AEJQD|1,1071.4455'
                                                b'45\ng(-Ac)m-AEJRD|1,1099.4880'
                                                b'78\ng(-Ac)m-AEJSD|1,1030.4189'
                                                b'95\ng(-Ac)m-AEJTD|1,1044.4346'
                                                b'45\ng(-Ac)m-AEJVD|1,1042.4553'
                                                b'81\ng(-Ac)m-AEJWD|1,1129.4662'
                                                b'80\ng(-Ac)m-AEJYD|1,1106.4502'
                                                b'96\ng(-Ac)m-AEJAE|1,1028.4397'
                                                b'31\ng(-Ac)m-AEJCE|1,1060.4118'
                                                b'02\ng(-Ac)m-AEJDE|1,1072.4295'
                                                b'60\ng(-Ac)m-AEJEE|1,1086.4452'
                                                b'10\ng(-Ac)m-AEJFE|1,1104.4710'
                                                b'31\ng(-Ac)m-AEJGE|1,1014.4240'
                                                b'81\ng(-Ac)m-AEJHE|1,1094.4615'
                                                b'29\ng(-Ac)m-AEJIE|1,1070.4866'
                                                b'81\ng(-Ac)m-AEJKE|1,1085.4975'
                                                b'80\ng(-Ac)m-AEJLE|1,1070.4866'
                                                b'81\ng(-Ac)m-AEJME|1,1088.4431'
                                                b'02\ng(-Ac)m-AEJNE|1,1071.4455'
                                                b'44\ng(-Ac)m-AEJPE|1,1054.4553'
                                                b'81\ng(-Ac)m-AEJQE|1,1085.4611'
                                                b'95\ng(-Ac)m-AEJRE|1,1113.5037'
                                                b'28\ng(-Ac)m-AEJSE|1,1044.4346'
                                                b'45\ng(-Ac)m-AEJTE|1,1058.4502'
                                                b'95\ng(-Ac)m-AEJVE|1,1056.4710'
                                                b'31\ng(-Ac)m-AEJWE|1,1143.4819'
                                                b'30\ng(-Ac)m-AEJYE|1,1120.4659'
                                                b'46\ng(-Ac)m-AEJAF|1,1046.4655'
                                                b'52\ng(-Ac)m-AEJCF|1,1078.4376'
                                                b'23\ng(-Ac)m-AEJDF|1,1090.4553'
                                                b'81\ng(-Ac)m-AEJEF|1,1104.4710'
                                                b'31\ng(-Ac)m-AEJFF|1,1122.4968'
                                                b'52\ng(-Ac)m-AEJGF|1,1032.4499'
                                                b'02\ng(-Ac)m-AEJHF|1,1112.4873'
                                                b'50\ng(-Ac)m-AEJIF|1,1088.5125'
                                                b'02\ng(-Ac)m-AEJKF|1,1103.5234'
                                                b'01\ng(-Ac)m-AEJLF|1,1088.5125'
                                                b'02\ng(-Ac)m-AEJMF|1,1106.4689'
                                                b'23\ng(-Ac)m-AEJNF|1,1089.4713'
                                                b'65\ng(-Ac)m-AEJPF|1,1072.4812'
                                                b'02\ng(-Ac)m-AEJQF|1,1103.4870'
                                                b'16\ng(-Ac)m-AEJRF|1,1131.5295'
                                                b'49\ng(-Ac)m-AEJSF|1,1062.4604'
                                                b'66\ng(-Ac)m-AEJTF|1,1076.4761'
                                                b'16\ng(-Ac)m-AEJVF|1,1074.4968'
                                                b'52\ng(-Ac)m-AEJWF|1,1161.5077'
                                                b'51\ng(-Ac)m-AEJYF|1,1138.4917'
                                                b'67\ng(-Ac)m-AEJAG|1,956.41860'
                                                b'2\ng(-Ac)m-AEJCG|1,988.390673'
                                                b'\ng(-Ac)m-AEJDG|1,1000.408431'
                                                b'\ng(-Ac)m-AEJEG|1,1014.424081'
                                                b'\ng(-Ac)m-AEJFG|1,1032.449902'
                                                b'\ng(-Ac)m-AEJGG|1,942.402'
                                                b'952\ng(-Ac)m-AEJHG|1,1022.440'
                                                b'400\ng(-Ac)m-AEJIG|1,998.4655'
                                                b'52\ng(-Ac)m-AEJKG|1,1013.4764'
                                                b'51\ng(-Ac)m-AEJLG|1,998.46555'
                                                b'2\ng(-Ac)m-AEJMG|1,1016.42197'
                                                b'3\ng(-Ac)m-AEJNG|1,999.424415'
                                                b'\ng(-Ac)m-AEJPG|1,982.434'
                                                b'252\ng(-Ac)m-AEJQG|1,1013.440'
                                                b'066\ng(-Ac)m-AEJRG|1,1041.482'
                                                b'599\ng(-Ac)m-AEJSG|1,972.4135'
                                                b'16\ng(-Ac)m-AEJTG|1,986.42916'
                                                b'6\ng(-Ac)m-AEJVG|1,984.449902'
                                                b'\ng(-Ac)m-AEJWG|1,1071.460801'
                                                b'\ng(-Ac)m-AEJYG|1,1048.444817'
                                                b'\ng(-Ac)m-AEJAH|1,1036.456050'
                                                b'\ng(-Ac)m-AEJCH|1,1068.428121'
                                                b'\ng(-Ac)m-AEJDH|1,1080.445879'
                                                b'\ng(-Ac)m-AEJEH|1,1094.461529'
                                                b'\ng(-Ac)m-AEJFH|1,1112.487350'
                                                b'\ng(-Ac)m-AEJGH|1,1022.440400'
                                                b'\ng(-Ac)m-AEJHH|1,1102.477848'
                                                b'\ng(-Ac)m-AEJIH|1,1078.503000'
                                                b'\ng(-Ac)m-AEJKH|1,1093.513899'
                                                b'\ng(-Ac)m-AEJLH|1,1078.503000'
                                                b'\ng(-Ac)m-AEJMH|1,1096.459421'
                                                b'\ng(-Ac)m-AEJNH|1,1079.461863'
                                                b'\ng(-Ac)m-AEJPH|1,1062.471700'
                                                b'\ng(-Ac)m-AEJQH|1,1093.477514'
                                                b'\ng(-Ac)m-AEJRH|1,1121.520047'
                                                b'\ng(-Ac)m-AEJSH|1,1052.450964'
                                                b'\ng(-Ac)m-AEJTH|1,1066.466614'
                                                b'\ng(-Ac)m-AEJVH|1,1064.487350'
                                                b'\ng(-Ac)m-AEJWH|1,1151.498249'
                                                b'\ng(-Ac)m-AEJYH|1,1128.482265'
                                                b'\ng(-Ac)m-AEJAI|1,1012.481202'
                                                b'\ng(-Ac)m-AEJCI|1,1044.453273'
                                                b'\ng(-Ac)m-AEJDI|1,1056.471031'
                                                b'\ng(-Ac)m-AEJEI|1,1070.486681'
                                                b'\ng(-Ac)m-AEJFI|1,1088.512502'
                                                b'\ng(-Ac)m-AEJGI|1,998.465'
                                                b'552\ng(-Ac)m-AEJHI|1,1078.503'
                                                b'000\ng(-Ac)m-AEJII|1,1054.528'
                                                b'152\ng(-Ac)m-AEJKI|1,1069.539'
                                                b'051\ng(-Ac)m-AEJLI|1,1054.528'
                                                b'152\ng(-Ac)m-AEJMI|1,1072.484'
                                                b'573\ng(-Ac)m-AEJNI|1,1055.487'
                                                b'015\ng(-Ac)m-AEJPI|1,1038.496'
                                                b'852\ng(-Ac)m-AEJQI|1,1069.502'
                                                b'666\ng(-Ac)m-AEJRI|1,1097.545'
                                                b'199\ng(-Ac)m-AEJSI|1,1028.476'
                                                b'116\ng(-Ac)m-AEJTI|1,1042.491'
                                                b'766\ng(-Ac)m-AEJVI|1,1040.512'
                                                b'502\ng(-Ac)m-AEJWI|1,1127.523'
                                                b'401\ng(-Ac)m-AEJYI|1,1104.507'
                                                b'417\ng(-Ac)m-AEJAK|1,1027.492'
                                                b'101\ng(-Ac)m-AEJCK|1,1059.464'
                                                b'172\ng(-Ac)m-AEJDK|1,1071.481'
                                                b'930\ng(-Ac)m-AEJEK|1,1085.497'
                                                b'580\ng(-Ac)m-AEJFK|1,1103.523'
                                                b'401\ng(-Ac)m-AEJGK|1,1013.476'
                                                b'451\ng(-Ac)m-AEJHK|1,1093.513'
                                                b'899\ng(-Ac)m-AEJIK|1,1069.539'
                                                b'051\ng(-Ac)m-AEJKK|1,1084.549'
                                                b'950\ng(-Ac)m-AEJLK|1,1069.539'
                                                b'051\ng(-Ac)m-AEJMK|1,1087.495'
                                                b'472\ng(-Ac)m-AEJNK|1,1070.497'
                                                b'914\ng(-Ac)m-AEJPK|1,1053.507'
                                                b'751\ng(-Ac)m-AEJQK|1,1084.513'
                                                b'565\ng(-Ac)m-AEJRK|1,1112.556'
                                                b'098\ng(-Ac)m-AEJSK|1,1043.487'
                                                b'015\ng(-Ac)m-AEJTK|1,1057.502'
                                                b'665\ng(-Ac)m-AEJVK|1,1055.523'
                                                b'401\ng(-Ac)m-AEJWK|1,1142.534'
                                                b'300\ng(-Ac)m-AEJYK|1,1119.518'
                                                b'316\ng(-Ac)m-AEJAL|1,1012.481'
                                                b'202\ng(-Ac)m-AEJCL|1,1044.453'
                                                b'273\ng(-Ac)m-AEJDL|1,1056.471'
                                                b'031\ng(-Ac)m-AEJEL|1,1070.486'
                                                b'681\ng(-Ac)m-AEJFL|1,1088.512'
                                                b'502\ng(-Ac)m-AEJGL|1,998.4655'
                                                b'52\ng(-Ac)m-AEJHL|1,1078.5030'
                                                b'00\ng(-Ac)m-AEJIL|1,1054.5281'
                                                b'52\ng(-Ac)m-AEJKL|1,1069.5390'
                                                b'51\ng(-Ac)m-AEJLL|1,1054.5281'
                                                b'52\ng(-Ac)m-AEJML|1,1072.4845'
                                                b'73\ng(-Ac)m-AEJNL|1,1055.4870'
                                                b'15\ng(-Ac)m-AEJPL|1,1038.4968'
                                                b'52\ng(-Ac)m-AEJQL|1,1069.5026'
                                                b'66\ng(-Ac)m-AEJRL|1,1097.5451'
                                                b'99\ng(-Ac)m-AEJSL|1,1028.4761'
                                                b'16\ng(-Ac)m-AEJTL|1,1042.4917'
                                                b'66\ng(-Ac)m-AEJVL|1,1040.5125'
                                                b'02\ng(-Ac)m-AEJWL|1,1127.5234'
                                                b'01\ng(-Ac)m-AEJYL|1,1104.5074'
                                                b'17\ng(-Ac)m-AEJAM|1,1030.4376'
                                                b'23\ng(-Ac)m-AEJCM|1,1062.4096'
                                                b'94\ng(-Ac)m-AEJDM|1,1074.4274'
                                                b'52\ng(-Ac)m-AEJEM|1,1088.4431'
                                                b'02\ng(-Ac)m-AEJFM|1,1106.4689'
                                                b'23\ng(-Ac)m-AEJGM|1,1016.4219'
                                                b'73\ng(-Ac)m-AEJHM|1,1096.4594'
                                                b'21\ng(-Ac)m-AEJIM|1,1072.4845'
                                                b'73\ng(-Ac)m-AEJKM|1,1087.4954'
                                                b'72\ng(-Ac)m-AEJLM|1,1072.4845'
                                                b'73\ng(-Ac)m-AEJMM|1,1090.4409'
                                                b'94\ng(-Ac)m-AEJNM|1,1073.4434'
                                                b'36\ng(-Ac)m-AEJPM|1,1056.4532'
                                                b'73\ng(-Ac)m-AEJQM|1,1087.4590'
                                                b'87\ng(-Ac)m-AEJRM|1,1115.5016'
                                                b'20\ng(-Ac)m-AEJSM|1,1046.4325'
                                                b'37\ng(-Ac)m-AEJTM|1,1060.4481'
                                                b'87\ng(-Ac)m-AEJVM|1,1058.4689'
                                                b'23\ng(-Ac)m-AEJWM|1,1145.4798'
                                                b'22\ng(-Ac)m-AEJYM|1,1122.4638'
                                                b'38\ng(-Ac)m-AEJAN|1,1013.4400'
                                                b'65\ng(-Ac)m-AEJCN|1,1045.4121'
                                                b'36\ng(-Ac)m-AEJDN|1,1057.4298'
                                                b'94\ng(-Ac)m-AEJEN|1,1071.4455'
                                                b'44\ng(-Ac)m-AEJFN|1,1089.4713'
                                                b'65\ng(-Ac)m-AEJGN|1,999.42441'
                                                b'5\ng(-Ac)m-AEJHN|1,1079.46186'
                                                b'3\ng(-Ac)m-AEJIN|1,1055.48701'
                                                b'5\ng(-Ac)m-AEJKN|1,1070.49791'
                                                b'4\ng(-Ac)m-AEJLN|1,1055.48701'
                                                b'5\ng(-Ac)m-AEJMN|1,1073.44343'
                                                b'6\ng(-Ac)m-AEJNN|1,1056.44587'
                                                b'8\ng(-Ac)m-AEJPN|1,1039.45571'
                                                b'5\ng(-Ac)m-AEJQN|1,1070.46152'
                                                b'9\ng(-Ac)m-AEJRN|1,1098.50406'
                                                b'2\ng(-Ac)m-AEJSN|1,1029.43497'
                                                b'9\ng(-Ac)m-AEJTN|1,1043.45062'
                                                b'9\ng(-Ac)m-AEJVN|1,1041.47136'
                                                b'5\ng(-Ac)m-AEJWN|1,1128.48226'
                                                b'4\ng(-Ac)m-AEJYN|1,1105.46628'
                                                b'0\ng(-Ac)m-AEJAP|1,996.449902'
                                                b'\ng(-Ac)m-AEJCP|1,1028.421973'
                                                b'\ng(-Ac)m-AEJDP|1,1040.439731'
                                                b'\ng(-Ac)m-AEJEP|1,1054.455381'
                                                b'\ng(-Ac)m-AEJFP|1,1072.481202'
                                                b'\ng(-Ac)m-AEJGP|1,982.434'
                                                b'252\ng(-Ac)m-AEJHP|1,1062.471'
                                                b'700\ng(-Ac)m-AEJIP|1,1038.496'
                                                b'852\ng(-Ac)m-AEJKP|1,1053.507'
                                                b'751\ng(-Ac)m-AEJLP|1,1038.496'
                                                b'852\ng(-Ac)m-AEJMP|1,1056.453'
                                                b'273\ng(-Ac)m-AEJNP|1,1039.455'
                                                b'715\ng(-Ac)m-AEJPP|1,1022.465'
                                                b'552\ng(-Ac)m-AEJQP|1,1053.471'
                                                b'366\ng(-Ac)m-AEJRP|1,1081.513'
                                                b'899\ng(-Ac)m-AEJSP|1,1012.444'
                                                b'816\ng(-Ac)m-AEJTP|1,1026.460'
                                                b'466\ng(-Ac)m-AEJVP|1,1024.481'
                                                b'202\ng(-Ac)m-AEJWP|1,1111.492'
                                                b'101\ng(-Ac)m-AEJYP|1,1088.476'
                                                b'117\ng(-Ac)m-AEJAQ|1,1027.455'
                                                b'716\ng(-Ac)m-AEJCQ|1,1059.427'
                                                b'787\ng(-Ac)m-AEJDQ|1,1071.445'
                                                b'545\ng(-Ac)m-AEJEQ|1,1085.461'
                                                b'195\ng(-Ac)m-AEJFQ|1,1103.487'
                                                b'016\ng(-Ac)m-AEJGQ|1,1013.440'
                                                b'066\ng(-Ac)m-AEJHQ|1,1093.477'
                                                b'514\ng(-Ac)m-AEJIQ|1,1069.502'
                                                b'666\ng(-Ac)m-AEJKQ|1,1084.513'
                                                b'565\ng(-Ac)m-AEJLQ|1,1069.502'
                                                b'666\ng(-Ac)m-AEJMQ|1,1087.459'
                                                b'087\ng(-Ac)m-AEJNQ|1,1070.461'
                                                b'529\ng(-Ac)m-AEJPQ|1,1053.471'
                                                b'366\ng(-Ac)m-AEJQQ|1,1084.477'
                                                b'180\ng(-Ac)m-AEJRQ|1,1112.519'
                                                b'713\ng(-Ac)m-AEJSQ|1,1043.450'
                                                b'630\ng(-Ac)m-AEJTQ|1,1057.466'
                                                b'280\ng(-Ac)m-AEJVQ|1,1055.487'
                                                b'016\ng(-Ac)m-AEJWQ|1,1142.497'
                                                b'915\ng(-Ac)m-AEJYQ|1,1119.481'
                                                b'931\ng(-Ac)m-AEJAR|1,1055.498'
                                                b'249\ng(-Ac)m-AEJCR|1,1087.470'
                                                b'320\ng(-Ac)m-AEJDR|1,1099.488'
                                                b'078\ng(-Ac)m-AEJER|1,1113.503'
                                                b'728\ng(-Ac)m-AEJFR|1,1131.529'
                                                b'549\ng(-Ac)m-AEJGR|1,1041.482'
                                                b'599\ng(-Ac)m-AEJHR|1,1121.520'
                                                b'047\ng(-Ac)m-AEJIR|1,1097.545'
                                                b'199\ng(-Ac)m-AEJKR|1,1112.556'
                                                b'098\ng(-Ac)m-AEJLR|1,1097.545'
                                                b'199\ng(-Ac)m-AEJMR|1,1115.501'
                                                b'620\ng(-Ac)m-AEJNR|1,1098.504'
                                                b'062\ng(-Ac)m-AEJPR|1,1081.513'
                                                b'899\ng(-Ac)m-AEJQR|1,1112.519'
                                                b'713\ng(-Ac)m-AEJRR|1,1140.562'
                                                b'246\ng(-Ac)m-AEJSR|1,1071.493'
                                                b'163\ng(-Ac)m-AEJTR|1,1085.508'
                                                b'813\ng(-Ac)m-AEJVR|1,1083.529'
                                                b'549\ng(-Ac)m-AEJWR|1,1170.540'
                                                b'448\ng(-Ac)m-AEJYR|1,1147.524'
                                                b'464\ng(-Ac)m-AEJAS|1,986.4291'
                                                b'66\ng(-Ac)m-AEJCS|1,1018.4012'
                                                b'37\ng(-Ac)m-AEJDS|1,1030.4189'
                                                b'95\ng(-Ac)m-AEJES|1,1044.4346'
                                                b'45\ng(-Ac)m-AEJFS|1,1062.4604'
                                                b'66\ng(-Ac)m-AEJGS|1,972.41351'
                                                b'6\ng(-Ac)m-AEJHS|1,1052.45096'
                                                b'4\ng(-Ac)m-AEJIS|1,1028.47611'
                                                b'6\ng(-Ac)m-AEJKS|1,1043.48701'
                                                b'5\ng(-Ac)m-AEJLS|1,1028.47611'
                                                b'6\ng(-Ac)m-AEJMS|1,1046.43253'
                                                b'7\ng(-Ac)m-AEJNS|1,1029.43497'
                                                b'9\ng(-Ac)m-AEJPS|1,1012.44481'
                                                b'6\ng(-Ac)m-AEJQS|1,1043.45063'
                                                b'0\ng(-Ac)m-AEJRS|1,1071.49316'
                                                b'3\ng(-Ac)m-AEJSS|1,1002.42408'
                                                b'0\ng(-Ac)m-AEJTS|1,1016.43973'
                                                b'0\ng(-Ac)m-AEJVS|1,1014.46046'
                                                b'6\ng(-Ac)m-AEJWS|1,1101.47136'
                                                b'5\ng(-Ac)m-AEJYS|1,1078.45538'
                                                b'1\ng(-Ac)m-AEJAT|1,1000.44481'
                                                b'6\ng(-Ac)m-AEJCT|1,1032.41688'
                                                b'7\ng(-Ac)m-AEJDT|1,1044.43464'
                                                b'5\ng(-Ac)m-AEJET|1,1058.45029'
                                                b'5\ng(-Ac)m-AEJFT|1,1076.47611'
                                                b'6\ng(-Ac)m-AEJGT|1,986.429166'
                                                b'\ng(-Ac)m-AEJHT|1,1066.466614'
                                                b'\ng(-Ac)m-AEJIT|1,1042.491766'
                                                b'\ng(-Ac)m-AEJKT|1,1057.502665'
                                                b'\ng(-Ac)m-AEJLT|1,1042.491766'
                                                b'\ng(-Ac)m-AEJMT|1,1060.448187'
                                                b'\ng(-Ac)m-AEJNT|1,1043.450629'
                                                b'\ng(-Ac)m-AEJPT|1,1026.460466'
                                                b'\ng(-Ac)m-AEJQT|1,1057.466280'
                                                b'\ng(-Ac)m-AEJRT|1,1085.508813'
                                                b'\ng(-Ac)m-AEJST|1,1016.439730'
                                                b'\ng(-Ac)m-AEJTT|1,1030.455380'
                                                b'\ng(-Ac)m-AEJVT|1,1028.476116'
                                                b'\ng(-Ac)m-AEJWT|1,1115.487015'
                                                b'\ng(-Ac)m-AEJYT|1,1092.471031'
                                                b'\ng(-Ac)m-AEJAV|1,998.465'
                                                b'552\ng(-Ac)m-AEJCV|1,1030.437'
                                                b'623\ng(-Ac)m-AEJDV|1,1042.455'
                                                b'381\ng(-Ac)m-AEJEV|1,1056.471'
                                                b'031\ng(-Ac)m-AEJFV|1,1074.496'
                                                b'852\ng(-Ac)m-AEJGV|1,984.4499'
                                                b'02\ng(-Ac)m-AEJHV|1,1064.4873'
                                                b'50\ng(-Ac)m-AEJIV|1,1040.5125'
                                                b'02\ng(-Ac)m-AEJKV|1,1055.5234'
                                                b'01\ng(-Ac)m-AEJLV|1,1040.5125'
                                                b'02\ng(-Ac)m-AEJMV|1,1058.4689'
                                                b'23\ng(-Ac)m-AEJNV|1,1041.4713'
                                                b'65\ng(-Ac)m-AEJPV|1,1024.4812'
                                                b'02\ng(-Ac)m-AEJQV|1,1055.4870'
                                                b'16\ng(-Ac)m-AEJRV|1,1083.5295'
                                                b'49\ng(-Ac)m-AEJSV|1,1014.4604'
                                                b'66\ng(-Ac)m-AEJTV|1,1028.4761'
                                                b'16\ng(-Ac)m-AEJVV|1,1026.4968'
                                                b'52\ng(-Ac)m-AEJWV|1,1113.5077'
                                                b'51\ng(-Ac)m-AEJYV|1,1090.4917'
                                                b'67\ng(-Ac)m-AEJAW|1,1085.4764'
                                                b'51\ng(-Ac)m-AEJCW|1,1117.4485'
                                                b'22\ng(-Ac)m-AEJDW|1,1129.4662'
                                                b'80\ng(-Ac)m-AEJEW|1,1143.4819'
                                                b'30\ng(-Ac)m-AEJFW|1,1161.5077'
                                                b'51\ng(-Ac)m-AEJGW|1,1071.4608'
                                                b'01\ng(-Ac)m-AEJHW|1,1151.4982'
                                                b'49\ng(-Ac)m-AEJIW|1,1127.5234'
                                                b'01\ng(-Ac)m-AEJKW|1,1142.5343'
                                                b'00\ng(-Ac)m-AEJLW|1,1127.5234'
                                                b'01\ng(-Ac)m-AEJMW|1,1145.4798'
                                                b'22\ng(-Ac)m-AEJNW|1,1128.4822'
                                                b'64\ng(-Ac)m-AEJPW|1,1111.4921'
                                                b'01\ng(-Ac)m-AEJQW|1,1142.4979'
                                                b'15\ng(-Ac)m-AEJRW|1,1170.5404'
                                                b'48\ng(-Ac)m-AEJSW|1,1101.4713'
                                                b'65\ng(-Ac)m-AEJTW|1,1115.4870'
                                                b'15\ng(-Ac)m-AEJVW|1,1113.5077'
                                                b'51\ng(-Ac)m-AEJWW|1,1200.5186'
                                                b'50\ng(-Ac)m-AEJYW|1,1177.5026'
                                                b'66\ng(-Ac)m-AEJAY|1,1062.4604'
                                                b'67\ng(-Ac)m-AEJCY|1,1094.4325'
                                                b'38\ng(-Ac)m-AEJDY|1,1106.4502'
                                                b'96\ng(-Ac)m-AEJEY|1,1120.4659'
                                                b'46\ng(-Ac)m-AEJFY|1,1138.4917'
                                                b'67\ng(-Ac)m-AEJGY|1,1048.4448'
                                                b'17\ng(-Ac)m-AEJHY|1,1128.4822'
                                                b'65\ng(-Ac)m-AEJIY|1,1104.5074'
                                                b'17\ng(-Ac)m-AEJKY|1,1119.5183'
                                                b'16\ng(-Ac)m-AEJLY|1,1104.5074'
                                                b'17\ng(-Ac)m-AEJMY|1,1122.4638'
                                                b'38\ng(-Ac)m-AEJNY|1,1105.4662'
                                                b'80\ng(-Ac)m-AEJPY|1,1088.4761'
                                                b'17\ng(-Ac)m-AEJQY|1,1119.4819'
                                                b'31\ng(-Ac)m-AEJRY|1,1147.5244'
                                                b'64\ng(-Ac)m-AEJSY|1,1078.4553'
                                                b'81\ng(-Ac)m-AEJTY|1,1092.4710'
                                                b'31\ng(-Ac)m-AEJVY|1,1090.4917'
                                                b'67\ng(-Ac)m-AEJWY|1,1177.5026'
                                                b'66\ng(-Ac)m-AEJYY|1,1154.4866'
                                                b'82\n',
                                     'Description': 'Contains di- tetra- and '
                                                    'hexasaccharides, '
                                                    'g(-Ac)m-A, g(-Ac)m-AE, '
                                                    'g(-Ac)m-AEJ, g(-Ac)m-AEJX '
                                                    'and all g(-Ac)m-AEJXX '
                                                    'structures, where X can '
                                                    'be any of the 20 '
                                                    'canonical amino acids.',
                                     'File': 'c_diff_monomers_complex.csv',
                                     'Name': 'Complex',
                                     'Species': 'Clostridium difficile'}],
    'Escherichia coli': [   {   'Content': b'Structure,Monoisotopicmass\ngm|0,'
                                           b'498.206090\ngm~gm|0,976.385965\ngm'
                                           b'~gm~gm|0,1454.565840\ngm-A|1,569.'
                                           b'243204\ngm-AE|1,698.285797\ngm-AEJ'
                                           b'|1,870.370589\ngm-AEJA|1,941.4077'
                                           b'03\ngm-AEJC|1,973.379774\ngm-AEJD|'
                                           b'1,985.397532\ngm-AEJE|1,999.41318'
                                           b'2\ngm-AEJF|1,1017.439003\ngm-AEJG|'
                                           b'1,927.392053\ngm-AEJH|1,1007.4295'
                                           b'01\ngm-AEJI|1,983.454653\ngm-AEJK|'
                                           b'1,998.465552\ngm-AEJM|1,1001.4110'
                                           b'74\ngm-AEJN|1,984.413516\ngm-AEJP|'
                                           b'1,967.423353\ngm-AEJQ|1,998.42916'
                                           b'7\ngm-AEJR|1,1026.471700\ngm-AEJS|'
                                           b'1,957.402617\ngm-AEJT|1,971.41826'
                                           b'7\ngm-AEJV|1,969.439003\ngm-AEJW|1'
                                           b',1056.449902\ngm-AEJY|1,1033.4339'
                                           b'18\ngm-AEJAA|1,1012.444817\ngm-AEJ'
                                           b'AC|1,1044.416888\ngm-AEJAD|1,1056'
                                           b'.434646\ngm-AEJAE|1,1070.450296\ng'
                                           b'm-AEJAF|1,1088.476117\ngm-AEJAG|1'
                                           b',998.429167\ngm-AEJAH|1,1078.4666'
                                           b'15\ngm-AEJAI|1,1054.491767\ngm-AEJ'
                                           b'AK|1,1069.502666\ngm-AEJAM|1,1072'
                                           b'.448188\ngm-AEJAN|1,1055.450630\ng'
                                           b'm-AEJAP|1,1038.460467\ngm-AEJAQ|1'
                                           b',1069.466281\ngm-AEJAR|1,1097.508'
                                           b'814\ngm-AEJAS|1,1028.439731\ngm-AE'
                                           b'JAT|1,1042.455381\ngm-AEJAV|1,104'
                                           b'0.476117\ngm-AEJAW|1,1127.487016\n'
                                           b'gm-AEJAY|1,1104.471032\n',
                                'Description': 'Contains di- tetra- and '
                                               'hexasaccharides, gm-A, gm-AE, '
                                               'gm-AEJ, gm-AEJX and all '
                                               'gm-AEJAX structures, where X '
                                               'can be any of the 20 canonical '
                                               'amino acids. In all '
                                               'structures, leucine is removed '
                                               'in favour of isoleucine.',
                                'File': 'e_coli_monomers_simple.csv',
                                'Name': 'Simple',
                                'Species': 'Escherichia coli'},
                            {   'Content': b'Structure,Monoisotopicmass\ngm|0,'
                                           b'498.206090\ngm~gm|0,976.385965\ngm'
                                           b'~gm~gm|0,1454.565840\ngm-A|1,569.'
                                           b'243204\ngm-AE|1,698.285797\ngm-AEJ'
                                           b'|1,870.370589\ngm-AEJA|1,941.4077'
                                           b'03\ngm-AEJC|1,973.379774\ngm-AEJD|'
                                           b'1,985.397532\ngm-AEJE|1,999.41318'
                                           b'2\ngm-AEJF|1,1017.439003\ngm-AEJG|'
                                           b'1,927.392053\ngm-AEJH|1,1007.4295'
                                           b'01\ngm-AEJI|1,983.454653\ngm-AEJK|'
                                           b'1,998.465552\ngm-AEJM|1,1001.4110'
                                           b'74\ngm-AEJN|1,984.413516\ngm-AEJP|'
                                           b'1,967.423353\ngm-AEJQ|1,998.42916'
                                           b'7\ngm-AEJR|1,1026.471700\ngm-AEJS|'
                                           b'1,957.402617\ngm-AEJT|1,971.41826'
                                           b'7\ngm-AEJV|1,969.439003\ngm-AEJW|1'
                                           b',1056.449902\ngm-AEJY|1,1033.4339'
                                           b'18\ngm-AEJAA|1,1012.444817\ngm-AEJ'
                                           b'AC|1,1044.416888\ngm-AEJAD|1,1056'
                                           b'.434646\ngm-AEJAE|1,1070.450296\ng'
                                           b'm-AEJAF|1,1088.476117\ngm-AEJAG|1'
                                           b',998.429167\ngm-AEJAH|1,1078.4666'
                                           b'15\ngm-AEJAI|1,1054.491767\ngm-AEJ'
                                           b'AK|1,1069.502666\ngm-AEJAM|1,1072'
                                           b'.448188\ngm-AEJAN|1,1055.450630\ng'
                                           b'm-AEJAP|1,1038.460467\ngm-AEJAQ|1'
                                           b',1069.466281\ngm-AEJAR|1,1097.508'
                                           b'814\ngm-AEJAS|1,1028.439731\ngm-AE'
                                           b'JAT|1,1042.455381\ngm-AEJAV|1,104'
                                           b'0.476117\ngm-AEJAW|1,1127.487016\n'
                                           b'gm-AEJAY|1,1104.471032\ngm-AEJCC|'
                                           b'1,1076.388959\ngm-AEJCD|1,1088.40'
                                           b'6717\ngm-AEJCE|1,1102.422367\ngm-A'
                                           b'EJCF|1,1120.448188\ngm-AEJCG|1,10'
                                           b'30.401238\ngm-AEJCH|1,1110.438686'
                                           b'\ngm-AEJCI|1,1086.463838\ngm-AEJCK'
                                           b'|1,1101.474737\ngm-AEJCM|1,1104.4'
                                           b'20259\ngm-AEJCN|1,1087.422701\ngm-'
                                           b'AEJCP|1,1070.432538\ngm-AEJCQ|1,1'
                                           b'101.438352\ngm-AEJCR|1,1129.48088'
                                           b'5\ngm-AEJCS|1,1060.411802\ngm-AEJC'
                                           b'T|1,1074.427452\ngm-AEJCV|1,1072.'
                                           b'448188\ngm-AEJCW|1,1159.459087\ngm'
                                           b'-AEJCY|1,1136.443103\ngm-AEJDD|1,'
                                           b'1100.424475\ngm-AEJDE|1,1114.4401'
                                           b'25\ngm-AEJDF|1,1132.465946\ngm-AEJ'
                                           b'DG|1,1042.418996\ngm-AEJDH|1,1122'
                                           b'.456444\ngm-AEJDI|1,1098.481596\ng'
                                           b'm-AEJDK|1,1113.492495\ngm-AEJDM|1'
                                           b',1116.438017\ngm-AEJDN|1,1099.440'
                                           b'459\ngm-AEJDP|1,1082.450296\ngm-AE'
                                           b'JDQ|1,1113.456110\ngm-AEJDR|1,114'
                                           b'1.498643\ngm-AEJDS|1,1072.429560\n'
                                           b'gm-AEJDT|1,1086.445210\ngm-AEJDV|'
                                           b'1,1084.465946\ngm-AEJDW|1,1171.47'
                                           b'6845\ngm-AEJDY|1,1148.460861\ngm-A'
                                           b'EJEE|1,1128.455775\ngm-AEJEF|1,11'
                                           b'46.481596\ngm-AEJEG|1,1056.434646'
                                           b'\ngm-AEJEH|1,1136.472094\ngm-AEJEI'
                                           b'|1,1112.497246\ngm-AEJEK|1,1127.5'
                                           b'08145\ngm-AEJEM|1,1130.453667\ngm-'
                                           b'AEJEN|1,1113.456109\ngm-AEJEP|1,1'
                                           b'096.465946\ngm-AEJEQ|1,1127.47176'
                                           b'0\ngm-AEJER|1,1155.514293\ngm-AEJE'
                                           b'S|1,1086.445210\ngm-AEJET|1,1100.'
                                           b'460860\ngm-AEJEV|1,1098.481596\ngm'
                                           b'-AEJEW|1,1185.492495\ngm-AEJEY|1,'
                                           b'1162.476511\ngm-AEJFF|1,1164.5074'
                                           b'17\ngm-AEJFG|1,1074.460467\ngm-AEJ'
                                           b'FH|1,1154.497915\ngm-AEJFI|1,1130'
                                           b'.523067\ngm-AEJFK|1,1145.533966\ng'
                                           b'm-AEJFM|1,1148.479488\ngm-AEJFN|1'
                                           b',1131.481930\ngm-AEJFP|1,1114.491'
                                           b'767\ngm-AEJFQ|1,1145.497581\ngm-AE'
                                           b'JFR|1,1173.540114\ngm-AEJFS|1,110'
                                           b'4.471031\ngm-AEJFT|1,1118.486681\n'
                                           b'gm-AEJFV|1,1116.507417\ngm-AEJFW|'
                                           b'1,1203.518316\ngm-AEJFY|1,1180.50'
                                           b'2332\ngm-AEJGG|1,984.413517\ngm-AE'
                                           b'JGH|1,1064.450965\ngm-AEJGI|1,104'
                                           b'0.476117\ngm-AEJGK|1,1055.487016\n'
                                           b'gm-AEJGM|1,1058.432538\ngm-AEJGN|'
                                           b'1,1041.434980\ngm-AEJGP|1,1024.44'
                                           b'4817\ngm-AEJGQ|1,1055.450631\ngm-A'
                                           b'EJGR|1,1083.493164\ngm-AEJGS|1,10'
                                           b'14.424081\ngm-AEJGT|1,1028.439731'
                                           b'\ngm-AEJGV|1,1026.460467\ngm-AEJGW'
                                           b'|1,1113.471366\ngm-AEJGY|1,1090.4'
                                           b'55382\ngm-AEJHH|1,1144.488413\ngm-'
                                           b'AEJHI|1,1120.513565\ngm-AEJHK|1,1'
                                           b'135.524464\ngm-AEJHM|1,1138.46998'
                                           b'6\ngm-AEJHN|1,1121.472428\ngm-AEJH'
                                           b'P|1,1104.482265\ngm-AEJHQ|1,1135.'
                                           b'488079\ngm-AEJHR|1,1163.530612\ngm'
                                           b'-AEJHS|1,1094.461529\ngm-AEJHT|1,'
                                           b'1108.477179\ngm-AEJHV|1,1106.4979'
                                           b'15\ngm-AEJHW|1,1193.508814\ngm-AEJ'
                                           b'HY|1,1170.492830\ngm-AEJII|1,1096'
                                           b'.538717\ngm-AEJIK|1,1111.549616\ng'
                                           b'm-AEJIM|1,1114.495138\ngm-AEJIN|1'
                                           b',1097.497580\ngm-AEJIP|1,1080.507'
                                           b'417\ngm-AEJIQ|1,1111.513231\ngm-AE'
                                           b'JIR|1,1139.555764\ngm-AEJIS|1,107'
                                           b'0.486681\ngm-AEJIT|1,1084.502331\n'
                                           b'gm-AEJIV|1,1082.523067\ngm-AEJIW|'
                                           b'1,1169.533966\ngm-AEJIY|1,1146.51'
                                           b'7982\ngm-AEJKK|1,1126.560515\ngm-A'
                                           b'EJKM|1,1129.506037\ngm-AEJKN|1,11'
                                           b'12.508479\ngm-AEJKP|1,1095.518316'
                                           b'\ngm-AEJKQ|1,1126.524130\ngm-AEJKR'
                                           b'|1,1154.566663\ngm-AEJKS|1,1085.4'
                                           b'97580\ngm-AEJKT|1,1099.513230\ngm-'
                                           b'AEJKV|1,1097.533966\ngm-AEJKW|1,1'
                                           b'184.544865\ngm-AEJKY|1,1161.52888'
                                           b'1\ngm-AEJMM|1,1132.451559\ngm-AEJM'
                                           b'N|1,1115.454001\ngm-AEJMP|1,1098.'
                                           b'463838\ngm-AEJMQ|1,1129.469652\ngm'
                                           b'-AEJMR|1,1157.512185\ngm-AEJMS|1,'
                                           b'1088.443102\ngm-AEJMT|1,1102.4587'
                                           b'52\ngm-AEJMV|1,1100.479488\ngm-AEJ'
                                           b'MW|1,1187.490387\ngm-AEJMY|1,1164'
                                           b'.474403\ngm-AEJNN|1,1098.456443\ng'
                                           b'm-AEJNP|1,1081.466280\ngm-AEJNQ|1'
                                           b',1112.472094\ngm-AEJNR|1,1140.514'
                                           b'627\ngm-AEJNS|1,1071.445544\ngm-AE'
                                           b'JNT|1,1085.461194\ngm-AEJNV|1,108'
                                           b'3.481930\ngm-AEJNW|1,1170.492829\n'
                                           b'gm-AEJNY|1,1147.476845\ngm-AEJPP|'
                                           b'1,1064.476117\ngm-AEJPQ|1,1095.48'
                                           b'1931\ngm-AEJPR|1,1123.524464\ngm-A'
                                           b'EJPS|1,1054.455381\ngm-AEJPT|1,10'
                                           b'68.471031\ngm-AEJPV|1,1066.491767'
                                           b'\ngm-AEJPW|1,1153.502666\ngm-AEJPY'
                                           b'|1,1130.486682\ngm-AEJQQ|1,1126.4'
                                           b'87745\ngm-AEJQR|1,1154.530278\ngm-'
                                           b'AEJQS|1,1085.461195\ngm-AEJQT|1,1'
                                           b'099.476845\ngm-AEJQV|1,1097.49758'
                                           b'1\ngm-AEJQW|1,1184.508480\ngm-AEJQ'
                                           b'Y|1,1161.492496\ngm-AEJRR|1,1182.'
                                           b'572811\ngm-AEJRS|1,1113.503728\ngm'
                                           b'-AEJRT|1,1127.519378\ngm-AEJRV|1,'
                                           b'1125.540114\ngm-AEJRW|1,1212.5510'
                                           b'13\ngm-AEJRY|1,1189.535029\ngm-AEJ'
                                           b'SS|1,1044.434645\ngm-AEJST|1,1058'
                                           b'.450295\ngm-AEJSV|1,1056.471031\ng'
                                           b'm-AEJSW|1,1143.481930\ngm-AEJSY|1'
                                           b',1120.465946\ngm-AEJTT|1,1072.465'
                                           b'945\ngm-AEJTV|1,1070.486681\ngm-AE'
                                           b'JTW|1,1157.497580\ngm-AEJTY|1,113'
                                           b'4.481596\ngm-AEJVV|1,1068.507417\n'
                                           b'gm-AEJVW|1,1155.518316\ngm-AEJVY|'
                                           b'1,1132.502332\ngm-AEJWW|1,1242.52'
                                           b'9215\ngm-AEJWY|1,1219.513231\ngm-A'
                                           b'EJYY|1,1196.497247\n',
                                'Description': 'Contains di- tetra- and '
                                               'hexasaccharides, gm-A, gm-AE, '
                                               'gm-AEJ, gm-AEJX and all '
                                               'gm-AEJXX structures, where X '
                                               'can be any of the 20 canonical '
                                               'amino acids; for gm-AEJXX '
                                               'structures, mass coincidences '
                                               'resulting from swapping the '
                                               'fourth and fifth stem residues '
                                               'have been removed. In all '
                                               'structures, leucine is removed '
                                               'in favour of isoleucine.',
                                'File': 'e_coli_monomers_non_redundant.csv',
                                'Name': 'Non-Redundant',
                                'Species': 'Escherichia coli'},
                            {   'Content': b'Structure,Monoisotopicmass\ngm|0,'
                                           b'498.206090\ngm~gm|0,976.385965\ngm'
                                           b'~gm~gm|0,1454.565840\ngm-A|1,569.'
                                           b'243204\ngm-AE|1,698.285797\ngm-AEJ'
                                           b'|1,870.370589\ngm-AEJA|1,941.4077'
                                           b'03\ngm-AEJC|1,973.379774\ngm-AEJD|'
                                           b'1,985.397532\ngm-AEJE|1,999.41318'
                                           b'2\ngm-AEJF|1,1017.439003\ngm-AEJG|'
                                           b'1,927.392053\ngm-AEJH|1,1007.4295'
                                           b'01\ngm-AEJI|1,983.454653\ngm-AEJK|'
                                           b'1,998.465552\ngm-AEJL|1,983.45465'
                                           b'3\ngm-AEJM|1,1001.411074\ngm-AEJN|'
                                           b'1,984.413516\ngm-AEJP|1,967.42335'
                                           b'3\ngm-AEJQ|1,998.429167\ngm-AEJR|1'
                                           b',1026.471700\ngm-AEJS|1,957.40261'
                                           b'7\ngm-AEJT|1,971.418267\ngm-AEJV|1'
                                           b',969.439003\ngm-AEJW|1,1056.44990'
                                           b'2\ngm-AEJY|1,1033.433918\ngm-AEJAA'
                                           b'|1,1012.444817\ngm-AEJAC|1,1044.4'
                                           b'16888\ngm-AEJAD|1,1056.434646\ngm-'
                                           b'AEJAE|1,1070.450296\ngm-AEJAF|1,1'
                                           b'088.476117\ngm-AEJAG|1,998.429167'
                                           b'\ngm-AEJAH|1,1078.466615\ngm-AEJAI'
                                           b'|1,1054.491767\ngm-AEJAK|1,1069.5'
                                           b'02666\ngm-AEJAL|1,1054.491767\ngm-'
                                           b'AEJAM|1,1072.448188\ngm-AEJAN|1,1'
                                           b'055.450630\ngm-AEJAP|1,1038.46046'
                                           b'7\ngm-AEJAQ|1,1069.466281\ngm-AEJA'
                                           b'R|1,1097.508814\ngm-AEJAS|1,1028.'
                                           b'439731\ngm-AEJAT|1,1042.455381\ngm'
                                           b'-AEJAV|1,1040.476117\ngm-AEJAW|1,'
                                           b'1127.487016\ngm-AEJAY|1,1104.4710'
                                           b'32\ngm-AEJCA|1,1044.416888\ngm-AEJ'
                                           b'CC|1,1076.388959\ngm-AEJCD|1,1088'
                                           b'.406717\ngm-AEJCE|1,1102.422367\ng'
                                           b'm-AEJCF|1,1120.448188\ngm-AEJCG|1'
                                           b',1030.401238\ngm-AEJCH|1,1110.438'
                                           b'686\ngm-AEJCI|1,1086.463838\ngm-AE'
                                           b'JCK|1,1101.474737\ngm-AEJCL|1,108'
                                           b'6.463838\ngm-AEJCM|1,1104.420259\n'
                                           b'gm-AEJCN|1,1087.422701\ngm-AEJCP|'
                                           b'1,1070.432538\ngm-AEJCQ|1,1101.43'
                                           b'8352\ngm-AEJCR|1,1129.480885\ngm-A'
                                           b'EJCS|1,1060.411802\ngm-AEJCT|1,10'
                                           b'74.427452\ngm-AEJCV|1,1072.448188'
                                           b'\ngm-AEJCW|1,1159.459087\ngm-AEJCY'
                                           b'|1,1136.443103\ngm-AEJDA|1,1056.4'
                                           b'34646\ngm-AEJDC|1,1088.406717\ngm-'
                                           b'AEJDD|1,1100.424475\ngm-AEJDE|1,1'
                                           b'114.440125\ngm-AEJDF|1,1132.46594'
                                           b'6\ngm-AEJDG|1,1042.418996\ngm-AEJD'
                                           b'H|1,1122.456444\ngm-AEJDI|1,1098.'
                                           b'481596\ngm-AEJDK|1,1113.492495\ngm'
                                           b'-AEJDL|1,1098.481596\ngm-AEJDM|1,'
                                           b'1116.438017\ngm-AEJDN|1,1099.4404'
                                           b'59\ngm-AEJDP|1,1082.450296\ngm-AEJ'
                                           b'DQ|1,1113.456110\ngm-AEJDR|1,1141'
                                           b'.498643\ngm-AEJDS|1,1072.429560\ng'
                                           b'm-AEJDT|1,1086.445210\ngm-AEJDV|1'
                                           b',1084.465946\ngm-AEJDW|1,1171.476'
                                           b'845\ngm-AEJDY|1,1148.460861\ngm-AE'
                                           b'JEA|1,1070.450296\ngm-AEJEC|1,110'
                                           b'2.422367\ngm-AEJED|1,1114.440125\n'
                                           b'gm-AEJEE|1,1128.455775\ngm-AEJEF|'
                                           b'1,1146.481596\ngm-AEJEG|1,1056.43'
                                           b'4646\ngm-AEJEH|1,1136.472094\ngm-A'
                                           b'EJEI|1,1112.497246\ngm-AEJEK|1,11'
                                           b'27.508145\ngm-AEJEL|1,1112.497246'
                                           b'\ngm-AEJEM|1,1130.453667\ngm-AEJEN'
                                           b'|1,1113.456109\ngm-AEJEP|1,1096.4'
                                           b'65946\ngm-AEJEQ|1,1127.471760\ngm-'
                                           b'AEJER|1,1155.514293\ngm-AEJES|1,1'
                                           b'086.445210\ngm-AEJET|1,1100.46086'
                                           b'0\ngm-AEJEV|1,1098.481596\ngm-AEJE'
                                           b'W|1,1185.492495\ngm-AEJEY|1,1162.'
                                           b'476511\ngm-AEJFA|1,1088.476117\ngm'
                                           b'-AEJFC|1,1120.448188\ngm-AEJFD|1,'
                                           b'1132.465946\ngm-AEJFE|1,1146.4815'
                                           b'96\ngm-AEJFF|1,1164.507417\ngm-AEJ'
                                           b'FG|1,1074.460467\ngm-AEJFH|1,1154'
                                           b'.497915\ngm-AEJFI|1,1130.523067\ng'
                                           b'm-AEJFK|1,1145.533966\ngm-AEJFL|1'
                                           b',1130.523067\ngm-AEJFM|1,1148.479'
                                           b'488\ngm-AEJFN|1,1131.481930\ngm-AE'
                                           b'JFP|1,1114.491767\ngm-AEJFQ|1,114'
                                           b'5.497581\ngm-AEJFR|1,1173.540114\n'
                                           b'gm-AEJFS|1,1104.471031\ngm-AEJFT|'
                                           b'1,1118.486681\ngm-AEJFV|1,1116.50'
                                           b'7417\ngm-AEJFW|1,1203.518316\ngm-A'
                                           b'EJFY|1,1180.502332\ngm-AEJGA|1,99'
                                           b'8.429167\ngm-AEJGC|1,1030.401238\n'
                                           b'gm-AEJGD|1,1042.418996\ngm-AEJGE|'
                                           b'1,1056.434646\ngm-AEJGF|1,1074.46'
                                           b'0467\ngm-AEJGG|1,984.413517\ngm-AE'
                                           b'JGH|1,1064.450965\ngm-AEJGI|1,104'
                                           b'0.476117\ngm-AEJGK|1,1055.487016\n'
                                           b'gm-AEJGL|1,1040.476117\ngm-AEJGM|'
                                           b'1,1058.432538\ngm-AEJGN|1,1041.43'
                                           b'4980\ngm-AEJGP|1,1024.444817\ngm-A'
                                           b'EJGQ|1,1055.450631\ngm-AEJGR|1,10'
                                           b'83.493164\ngm-AEJGS|1,1014.424081'
                                           b'\ngm-AEJGT|1,1028.439731\ngm-AEJGV'
                                           b'|1,1026.460467\ngm-AEJGW|1,1113.4'
                                           b'71366\ngm-AEJGY|1,1090.455382\ngm-'
                                           b'AEJHA|1,1078.466615\ngm-AEJHC|1,1'
                                           b'110.438686\ngm-AEJHD|1,1122.45644'
                                           b'4\ngm-AEJHE|1,1136.472094\ngm-AEJH'
                                           b'F|1,1154.497915\ngm-AEJHG|1,1064.'
                                           b'450965\ngm-AEJHH|1,1144.488413\ngm'
                                           b'-AEJHI|1,1120.513565\ngm-AEJHK|1,'
                                           b'1135.524464\ngm-AEJHL|1,1120.5135'
                                           b'65\ngm-AEJHM|1,1138.469986\ngm-AEJ'
                                           b'HN|1,1121.472428\ngm-AEJHP|1,1104'
                                           b'.482265\ngm-AEJHQ|1,1135.488079\ng'
                                           b'm-AEJHR|1,1163.530612\ngm-AEJHS|1'
                                           b',1094.461529\ngm-AEJHT|1,1108.477'
                                           b'179\ngm-AEJHV|1,1106.497915\ngm-AE'
                                           b'JHW|1,1193.508814\ngm-AEJHY|1,117'
                                           b'0.492830\ngm-AEJIA|1,1054.491767\n'
                                           b'gm-AEJIC|1,1086.463838\ngm-AEJID|'
                                           b'1,1098.481596\ngm-AEJIE|1,1112.49'
                                           b'7246\ngm-AEJIF|1,1130.523067\ngm-A'
                                           b'EJIG|1,1040.476117\ngm-AEJIH|1,11'
                                           b'20.513565\ngm-AEJII|1,1096.538717'
                                           b'\ngm-AEJIK|1,1111.549616\ngm-AEJIL'
                                           b'|1,1096.538717\ngm-AEJIM|1,1114.4'
                                           b'95138\ngm-AEJIN|1,1097.497580\ngm-'
                                           b'AEJIP|1,1080.507417\ngm-AEJIQ|1,1'
                                           b'111.513231\ngm-AEJIR|1,1139.55576'
                                           b'4\ngm-AEJIS|1,1070.486681\ngm-AEJI'
                                           b'T|1,1084.502331\ngm-AEJIV|1,1082.'
                                           b'523067\ngm-AEJIW|1,1169.533966\ngm'
                                           b'-AEJIY|1,1146.517982\ngm-AEJKA|1,'
                                           b'1069.502666\ngm-AEJKC|1,1101.4747'
                                           b'37\ngm-AEJKD|1,1113.492495\ngm-AEJ'
                                           b'KE|1,1127.508145\ngm-AEJKF|1,1145'
                                           b'.533966\ngm-AEJKG|1,1055.487016\ng'
                                           b'm-AEJKH|1,1135.524464\ngm-AEJKI|1'
                                           b',1111.549616\ngm-AEJKK|1,1126.560'
                                           b'515\ngm-AEJKL|1,1111.549616\ngm-AE'
                                           b'JKM|1,1129.506037\ngm-AEJKN|1,111'
                                           b'2.508479\ngm-AEJKP|1,1095.518316\n'
                                           b'gm-AEJKQ|1,1126.524130\ngm-AEJKR|'
                                           b'1,1154.566663\ngm-AEJKS|1,1085.49'
                                           b'7580\ngm-AEJKT|1,1099.513230\ngm-A'
                                           b'EJKV|1,1097.533966\ngm-AEJKW|1,11'
                                           b'84.544865\ngm-AEJKY|1,1161.528881'
                                           b'\ngm-AEJLA|1,1054.491767\ngm-AEJLC'
                                           b'|1,1086.463838\ngm-AEJLD|1,1098.4'
                                           b'81596\ngm-AEJLE|1,1112.497246\ngm-'
                                           b'AEJLF|1,1130.523067\ngm-AEJLG|1,1'
                                           b'040.476117\ngm-AEJLH|1,1120.51356'
                                           b'5\ngm-AEJLI|1,1096.538717\ngm-AEJL'
                                           b'K|1,1111.549616\ngm-AEJLL|1,1096.'
                                           b'538717\ngm-AEJLM|1,1114.495138\ngm'
                                           b'-AEJLN|1,1097.497580\ngm-AEJLP|1,'
                                           b'1080.507417\ngm-AEJLQ|1,1111.5132'
                                           b'31\ngm-AEJLR|1,1139.555764\ngm-AEJ'
                                           b'LS|1,1070.486681\ngm-AEJLT|1,1084'
                                           b'.502331\ngm-AEJLV|1,1082.523067\ng'
                                           b'm-AEJLW|1,1169.533966\ngm-AEJLY|1'
                                           b',1146.517982\ngm-AEJMA|1,1072.448'
                                           b'188\ngm-AEJMC|1,1104.420259\ngm-AE'
                                           b'JMD|1,1116.438017\ngm-AEJME|1,113'
                                           b'0.453667\ngm-AEJMF|1,1148.479488\n'
                                           b'gm-AEJMG|1,1058.432538\ngm-AEJMH|'
                                           b'1,1138.469986\ngm-AEJMI|1,1114.49'
                                           b'5138\ngm-AEJMK|1,1129.506037\ngm-A'
                                           b'EJML|1,1114.495138\ngm-AEJMM|1,11'
                                           b'32.451559\ngm-AEJMN|1,1115.454001'
                                           b'\ngm-AEJMP|1,1098.463838\ngm-AEJMQ'
                                           b'|1,1129.469652\ngm-AEJMR|1,1157.5'
                                           b'12185\ngm-AEJMS|1,1088.443102\ngm-'
                                           b'AEJMT|1,1102.458752\ngm-AEJMV|1,1'
                                           b'100.479488\ngm-AEJMW|1,1187.49038'
                                           b'7\ngm-AEJMY|1,1164.474403\ngm-AEJN'
                                           b'A|1,1055.450630\ngm-AEJNC|1,1087.'
                                           b'422701\ngm-AEJND|1,1099.440459\ngm'
                                           b'-AEJNE|1,1113.456109\ngm-AEJNF|1,'
                                           b'1131.481930\ngm-AEJNG|1,1041.4349'
                                           b'80\ngm-AEJNH|1,1121.472428\ngm-AEJ'
                                           b'NI|1,1097.497580\ngm-AEJNK|1,1112'
                                           b'.508479\ngm-AEJNL|1,1097.497580\ng'
                                           b'm-AEJNM|1,1115.454001\ngm-AEJNN|1'
                                           b',1098.456443\ngm-AEJNP|1,1081.466'
                                           b'280\ngm-AEJNQ|1,1112.472094\ngm-AE'
                                           b'JNR|1,1140.514627\ngm-AEJNS|1,107'
                                           b'1.445544\ngm-AEJNT|1,1085.461194\n'
                                           b'gm-AEJNV|1,1083.481930\ngm-AEJNW|'
                                           b'1,1170.492829\ngm-AEJNY|1,1147.47'
                                           b'6845\ngm-AEJPA|1,1038.460467\ngm-A'
                                           b'EJPC|1,1070.432538\ngm-AEJPD|1,10'
                                           b'82.450296\ngm-AEJPE|1,1096.465946'
                                           b'\ngm-AEJPF|1,1114.491767\ngm-AEJPG'
                                           b'|1,1024.444817\ngm-AEJPH|1,1104.4'
                                           b'82265\ngm-AEJPI|1,1080.507417\ngm-'
                                           b'AEJPK|1,1095.518316\ngm-AEJPL|1,1'
                                           b'080.507417\ngm-AEJPM|1,1098.46383'
                                           b'8\ngm-AEJPN|1,1081.466280\ngm-AEJP'
                                           b'P|1,1064.476117\ngm-AEJPQ|1,1095.'
                                           b'481931\ngm-AEJPR|1,1123.524464\ngm'
                                           b'-AEJPS|1,1054.455381\ngm-AEJPT|1,'
                                           b'1068.471031\ngm-AEJPV|1,1066.4917'
                                           b'67\ngm-AEJPW|1,1153.502666\ngm-AEJ'
                                           b'PY|1,1130.486682\ngm-AEJQA|1,1069'
                                           b'.466281\ngm-AEJQC|1,1101.438352\ng'
                                           b'm-AEJQD|1,1113.456110\ngm-AEJQE|1'
                                           b',1127.471760\ngm-AEJQF|1,1145.497'
                                           b'581\ngm-AEJQG|1,1055.450631\ngm-AE'
                                           b'JQH|1,1135.488079\ngm-AEJQI|1,111'
                                           b'1.513231\ngm-AEJQK|1,1126.524130\n'
                                           b'gm-AEJQL|1,1111.513231\ngm-AEJQM|'
                                           b'1,1129.469652\ngm-AEJQN|1,1112.47'
                                           b'2094\ngm-AEJQP|1,1095.481931\ngm-A'
                                           b'EJQQ|1,1126.487745\ngm-AEJQR|1,11'
                                           b'54.530278\ngm-AEJQS|1,1085.461195'
                                           b'\ngm-AEJQT|1,1099.476845\ngm-AEJQV'
                                           b'|1,1097.497581\ngm-AEJQW|1,1184.5'
                                           b'08480\ngm-AEJQY|1,1161.492496\ngm-'
                                           b'AEJRA|1,1097.508814\ngm-AEJRC|1,1'
                                           b'129.480885\ngm-AEJRD|1,1141.49864'
                                           b'3\ngm-AEJRE|1,1155.514293\ngm-AEJR'
                                           b'F|1,1173.540114\ngm-AEJRG|1,1083.'
                                           b'493164\ngm-AEJRH|1,1163.530612\ngm'
                                           b'-AEJRI|1,1139.555764\ngm-AEJRK|1,'
                                           b'1154.566663\ngm-AEJRL|1,1139.5557'
                                           b'64\ngm-AEJRM|1,1157.512185\ngm-AEJ'
                                           b'RN|1,1140.514627\ngm-AEJRP|1,1123'
                                           b'.524464\ngm-AEJRQ|1,1154.530278\ng'
                                           b'm-AEJRR|1,1182.572811\ngm-AEJRS|1'
                                           b',1113.503728\ngm-AEJRT|1,1127.519'
                                           b'378\ngm-AEJRV|1,1125.540114\ngm-AE'
                                           b'JRW|1,1212.551013\ngm-AEJRY|1,118'
                                           b'9.535029\ngm-AEJSA|1,1028.439731\n'
                                           b'gm-AEJSC|1,1060.411802\ngm-AEJSD|'
                                           b'1,1072.429560\ngm-AEJSE|1,1086.44'
                                           b'5210\ngm-AEJSF|1,1104.471031\ngm-A'
                                           b'EJSG|1,1014.424081\ngm-AEJSH|1,10'
                                           b'94.461529\ngm-AEJSI|1,1070.486681'
                                           b'\ngm-AEJSK|1,1085.497580\ngm-AEJSL'
                                           b'|1,1070.486681\ngm-AEJSM|1,1088.4'
                                           b'43102\ngm-AEJSN|1,1071.445544\ngm-'
                                           b'AEJSP|1,1054.455381\ngm-AEJSQ|1,1'
                                           b'085.461195\ngm-AEJSR|1,1113.50372'
                                           b'8\ngm-AEJSS|1,1044.434645\ngm-AEJS'
                                           b'T|1,1058.450295\ngm-AEJSV|1,1056.'
                                           b'471031\ngm-AEJSW|1,1143.481930\ngm'
                                           b'-AEJSY|1,1120.465946\ngm-AEJTA|1,'
                                           b'1042.455381\ngm-AEJTC|1,1074.4274'
                                           b'52\ngm-AEJTD|1,1086.445210\ngm-AEJ'
                                           b'TE|1,1100.460860\ngm-AEJTF|1,1118'
                                           b'.486681\ngm-AEJTG|1,1028.439731\ng'
                                           b'm-AEJTH|1,1108.477179\ngm-AEJTI|1'
                                           b',1084.502331\ngm-AEJTK|1,1099.513'
                                           b'230\ngm-AEJTL|1,1084.502331\ngm-AE'
                                           b'JTM|1,1102.458752\ngm-AEJTN|1,108'
                                           b'5.461194\ngm-AEJTP|1,1068.471031\n'
                                           b'gm-AEJTQ|1,1099.476845\ngm-AEJTR|'
                                           b'1,1127.519378\ngm-AEJTS|1,1058.45'
                                           b'0295\ngm-AEJTT|1,1072.465945\ngm-A'
                                           b'EJTV|1,1070.486681\ngm-AEJTW|1,11'
                                           b'57.497580\ngm-AEJTY|1,1134.481596'
                                           b'\ngm-AEJVA|1,1040.476117\ngm-AEJVC'
                                           b'|1,1072.448188\ngm-AEJVD|1,1084.4'
                                           b'65946\ngm-AEJVE|1,1098.481596\ngm-'
                                           b'AEJVF|1,1116.507417\ngm-AEJVG|1,1'
                                           b'026.460467\ngm-AEJVH|1,1106.49791'
                                           b'5\ngm-AEJVI|1,1082.523067\ngm-AEJV'
                                           b'K|1,1097.533966\ngm-AEJVL|1,1082.'
                                           b'523067\ngm-AEJVM|1,1100.479488\ngm'
                                           b'-AEJVN|1,1083.481930\ngm-AEJVP|1,'
                                           b'1066.491767\ngm-AEJVQ|1,1097.4975'
                                           b'81\ngm-AEJVR|1,1125.540114\ngm-AEJ'
                                           b'VS|1,1056.471031\ngm-AEJVT|1,1070'
                                           b'.486681\ngm-AEJVV|1,1068.507417\ng'
                                           b'm-AEJVW|1,1155.518316\ngm-AEJVY|1'
                                           b',1132.502332\ngm-AEJWA|1,1127.487'
                                           b'016\ngm-AEJWC|1,1159.459087\ngm-AE'
                                           b'JWD|1,1171.476845\ngm-AEJWE|1,118'
                                           b'5.492495\ngm-AEJWF|1,1203.518316\n'
                                           b'gm-AEJWG|1,1113.471366\ngm-AEJWH|'
                                           b'1,1193.508814\ngm-AEJWI|1,1169.53'
                                           b'3966\ngm-AEJWK|1,1184.544865\ngm-A'
                                           b'EJWL|1,1169.533966\ngm-AEJWM|1,11'
                                           b'87.490387\ngm-AEJWN|1,1170.492829'
                                           b'\ngm-AEJWP|1,1153.502666\ngm-AEJWQ'
                                           b'|1,1184.508480\ngm-AEJWR|1,1212.5'
                                           b'51013\ngm-AEJWS|1,1143.481930\ngm-'
                                           b'AEJWT|1,1157.497580\ngm-AEJWV|1,1'
                                           b'155.518316\ngm-AEJWW|1,1242.52921'
                                           b'5\ngm-AEJWY|1,1219.513231\ngm-AEJY'
                                           b'A|1,1104.471032\ngm-AEJYC|1,1136.'
                                           b'443103\ngm-AEJYD|1,1148.460861\ngm'
                                           b'-AEJYE|1,1162.476511\ngm-AEJYF|1,'
                                           b'1180.502332\ngm-AEJYG|1,1090.4553'
                                           b'82\ngm-AEJYH|1,1170.492830\ngm-AEJ'
                                           b'YI|1,1146.517982\ngm-AEJYK|1,1161'
                                           b'.528881\ngm-AEJYL|1,1146.517982\ng'
                                           b'm-AEJYM|1,1164.474403\ngm-AEJYN|1'
                                           b',1147.476845\ngm-AEJYP|1,1130.486'
                                           b'682\ngm-AEJYQ|1,1161.492496\ngm-AE'
                                           b'JYR|1,1189.535029\ngm-AEJYS|1,112'
                                           b'0.465946\ngm-AEJYT|1,1134.481596\n'
                                           b'gm-AEJYV|1,1132.502332\ngm-AEJYW|'
                                           b'1,1219.513231\ngm-AEJYY|1,1196.49'
                                           b'7247\n',
                                'Description': 'Contains di- tetra- and '
                                               'hexasaccharides, gm-A, gm-AE, '
                                               'gm-AEJ, gm-AEJX and all '
                                               'gm-AEJXX structures, where X '
                                               'can be any of the 20 canonical '
                                               'amino acids.',
                                'File': 'e_coli_monomers_complex.csv',
                                'Name': 'Complex',
                                'Species': 'Escherichia coli'}]}
```

Reply:
Oh, and the reason for this indexed-by-species structure has to do with how it's used in the UI! Each species forms a drop-down with the list of libraries inside:
![image](https://github.com/Mesnage-Org/pgfinder/assets/6251883/03eaf8d0-404b-4a5b-b544-fc1e3fa3aeb0)

Reply:
The data of structures/masses look like they've got mangled in the process of importing, but that may be my misunderstanding of the pretty printing.

I don't have masses of time to spend on this but started and decided to finish...

The following gives a nested dictionary with the following levels of nesting...

1. `species` (`Escherichia coli` / `Clostridium difficile`)
2. `names` (`Simple` / `Non-Redundant` / `Complex`). 
3. The "data" which consists of `file`, `description`, `structures
4.  `structures` is the actual data, the keys are the actual structure and values are the  masses (easy to flip if required).

This seems a bit cleaner to me in terms of structure and naturally lends itself to selecting first a sepcies, then the name and then the information you want about it (be that the `description` or the `structures` (`file` is only really used in reading the masses themselves, users wouldn't ever need to see it). A cleaner way of getting the data into a dictionary and also of then extracting data.

It might require a bit more work to get this working with the existing code/structure but the above is (to me) easier to read (but then I wrote it so I have that advantage this time round :wink: ).

```python
import csv
from pathlib import Path


BASE_DIR = Path().cwd()
DATA_DIR = BASE_DIR / "pgfinder" / "data"
MASSES_DIR = BASE_DIR / "pgfinder" / "masses"

index_file = DATA_DIR / "index.csv"

mass_data = {}
count = 1
with index_file.open() as csvfile:
    reader = csv.DictReader(csvfile)
    for row in reader:
        # If this is the first time we've seen this species we create a new entry with a nested dictionary indexed by
        # name
        if row["Species"] not in mass_data.keys():
            mass_data[row["Species"]] = {row["Name"]: {"description": row["Description"], "file": row["File"]}}
        # If we have already seen this species then we have a new Name so we append the description and file. We don't
        # need to include the Species as its information held at a higher level.
        else:
            mass_data[row["Species"]][row["Name"]] = {"description": row["Description"], "file": row["File"]}
# This gives us a nested dictionary indexed first by species with the different types of structures and then by the
# Name(/type of mass) that contains the file we want to read.
print(mass_data)


# We can now read each file of the actual masses, using the filename from index.csv whc

# Loop over species
for species, names in mass_data.items():
    # Loop over each name and get the data associated with that, this includes the file and we have set where that is
    # saved further above in MASSES_DIR so we can open and read that
    for name, data in names.items():
        mass_file = MASSES_DIR / data["file"]
        with mass_file.open() as csvfile:
            reader = csv.DictReader(csvfile)
            # Save the structures we're reading into a nested dictionary, with keys being the structure and values the
            # masses. 
            data["structures"] = {}
            for row in reader:
                data["structures"][row["Structure"]] = row["Monoisotopicmass"]
```

If you want to print the keys you can use recursion...

```python
def print_keys(dic):
    for key, value in dic.items():
        print(key)
        if isinstance(value, dict):
            print_keys(value)
print_keys(mass_data)
```

Reply:
Of course, if all the data were saved in JSON in the first instance this wouldn't be an issue we could just read JSON straight into a dictionary! :laughing:

Reply:
Hi @ns-rse ! This looks good and the nesting does feel more logical! It's currently only used by the UI though, so I'll have to see how this sort of structure fits in there. For example, since the mass library widget also allows for a custom library to be uploaded, I've waited to parses the actually structure → mass mappings until the analysis is run.

The sad and annoying bit is that whatever data structure we return here needs to be parsed apart again on the JS side of things, and I've found JS significantly more cumbersome to do these data transformations in than Python. I'll see how to best slot things together! Maybe the long-term ideal is a shim-package that can transform the PGFinder information into something UI friendly without cluttering up this actual package.

I'll keep you updated and thanks for the continued input!

Reply:
> I'll have to see how this sort of structure fits in there. For example, since the mass library widget also allows for a custom library to be uploaded, I've waited to parses the actually structure → mass mappings until the analysis is run.

I was only half joking in the above comment about storing this data structure as JSON, it may remove overhead in terms of reading either from the defaults or custom libraries.

Reply:
> This looks good and the nesting does feel more logical!

It's a subtle difference compared to what you had but I'm glad you think it makes more sense.

Reply:
Well I was thinking of modifying _all_ files to JSON but I can see how that might cause problems if additional data were to be added.

I don't see any problem moving GUI components here, they could go in their own directory, but need to make sure they're included in the package for building by updating `pyproject.toml` to include the directory.

Reply:
Index is JSON now, with mass files still easy-to-edit CSV. Now the CSV is only loaded when it's needed by PGFinder so we don't need to store all of the mass libraries in a single object in memory.

Pyodide complained a bit about a `pgfinder_gui` folder, so I gave up and used `pgfinder/gui` instead for the `shim.py` (everything needed by the web GUI) and `internal.py` (parts of the code that are GUI-specific but not called directly from any Javascript).

Overall very happy with the relatively clean solution this ended up being!